jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java

```diff
@@ -709,6 +709,30 @@ public Object optimizedClone(Object srcObj) throws CloneNotSupportedException {
 	}
 
 	/**
+	 * Get class initialize status flag. Calls to this method will be recognized and optimized by the JIT.
+	 * @parm defc
+	 *          The class whose initialize status is desired.
+	 * @return
+	 *          initializeStatus from J9Class.
+	 */
+	public final int getClassInitializeStatus(Class<?> defc) {
```

Comment:
There's actually potentially a cheaper way to check this - see the code in J9VMInternals.java:
```
        /**
         * Initialize a Class. See chapter
         * 2.17.5 of the JVM Specification (2nd ed)
         */
        static void initialize(Class<?> clazz) {
                /* initializationLock == null means initialization successfully completed */
                if (null != clazz.initializationLock) {
                        Unsafe localUnsafe = unsafe;
                        if (null == localUnsafe) {
                                localUnsafe = unsafe = Unsafe.getUnsafe();
                        }
                        localUnsafe.ensureClassInitialized(clazz);
                }
        }
```
No unsafe required.  A small change to VMAccess.java to expose the method as i.e. `isInitialized(Class<?> )` and then a call such as
```
VM.getVMLangAccess().isInitialized(defc)
```

Is that better from a JIT perspective?

Reply:
I can understand that this approach makes interpreter faster, however, it does not make difference to the JIT. When JIT sees the call to `getClassInitializeStatus`, it will convert it to a load of the `J9Class` from `defc` followed by a load of `initializeStatus` from the `J9Class` without looking into its java implementation, so there is no unsafe involved in the jitted code.

Reply:
I prefer a design that implements this in Java using an approach like:
```
class java.lang.Class {
...
/* New method that checks the initializedLock has been nulled out or not */
boolean isInitialized() {
  return initializationLock == null;
}
...
}

interface com.ibm.oti.vm.VMLangAccess {
/* new method */
boolean isClassInitialized(Class<?> aClass);
}

class java.lang.VMAccess implements VMLangAccess {
/* New method that provides the implementation */
boolean isClassInitialized(Class<?> aClass) {
  return aClass.isInitialized();
}
}
```

This approach seems cleaner to me as it avoids hardcoding VM Internals into the Java source.  Using the `null == clazz.initializationLock` check provides reasonable Java code, embeds it in an API that can be generally used, and makes it easier to modify the code in the future (no worries about how the / where the INITIALIZE_STATUS field lives or is changed).

We have also very occasionally seen cases were the JITHelpers methods are not compiled and they've been interpreted.  This makes the method cheap in those very unlikely to occur situations.

Is the JIT's recognized method still required with this straightforward of an implementation?  Does it also allow removing the special case from the JIT?

Reply:
Thanks for your thought into a cleaner design. Your approach is cleaner in terms of java code but it's hard/impossible for the JIT to optimize it.

First of all, I need to make a correction to my answer to your first question, i.e. is the following approach better from a JIT perspective?
```
VM.getVMLangAccess().isInitialized(defc)
```
I'm sorry that I had misunderstood your question last week. I thought you mean we're going to keep the JITHelpers method `getClassInitializeStatus` and only change its java implementation. If so, there is no difference to the JIT because we'll always recognize the method and turn it into two loads. So now the question is whether replacing `getClassInitializeStatus` with `isClassInitialized` better for the JIT. 

The reason of having `getClassInitializeStatus` instead of `isClassInitialized` in the first place is `initializeStatus` in `J9Class` encodes several values, and only one value implys a success in the class initialization process, the JIT can't simply generate load of `initializeStatus` for question `isClassInitialized`. So there is a difference in the JIT if we were to change the design to your approach, and the difference is the JIT can't fold the check.

Another question is: can we still use `isClassInitialized` but check `Class.initializationLock` ?
The answer is the JIT can't optimize `Class.initializationLock` because :

1) There is a chain of calls to do the check, to exposed the `initializationLock`, the JIT has to inline 4 methods, among which one is interface call. This is hard for the JIT and has the following issues:
 a) Consumes the inliner budget
 b) The inlining of the interface call needs to be guarded, resulting in runtime overhead.

2) The essential problem is that `Class.initializationLock` cannot be folded because it is not a final field. Whereas the whole purpose of this PR is to fold the class initialization check to make thunkArchetype and VarHandle faster.

Another thing is that as far as I'm concerned, we're not hardcoding VM Internals into the Java source, VM fills in the values of those fields, right? And we're already doing that for lots of JITHelpers methods

I hope my answer clears your question.

Reply:
Thanks @liqunl.  That makes a lot of sense.  I withdraw my objection =)